File: src/main/scala/com/pingcap/tispark/TiSessionCache.scala
/*
 * Copyright 2017 PingCAP, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pingcap.tispark

import java.util.HashMap

import com.pingcap.tikv.{TiConfiguration, TiSession}
import org.apache.spark.SparkContext

object TiSessionCache {
  private val sessionCache: HashMap[String, TiSession] = new HashMap[String, TiSession]()

  def getSession(appId: String, conf: TiConfiguration): TiSession = this.synchronized {
    val session = sessionCache.get(appId)
    if (session == null) {
      val newSession = TiSession.create(conf)
      sessionCache.put(appId, newSession)
      newSession
    } else {
      session
    }
  }

  def putSession(appId: String, session: TiSession): Unit = this.synchronized {
    if (session != null && appId != null && !appId.isEmpty) {
      sessionCache.put(appId, session)
    }
  }
}

File: tikv-client/src/main/java/com/pingcap/tikv/TiSessionCache.java
/*
 * Copyright 2019 PingCAP, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pingcap.tikv;

import java.util.HashMap;
import java.util.Map;

public class TiSessionCache {
  private static Map<String, TiSession> sessionCachedMap = new HashMap<>();

  // Since we create session as singleton now, configuration change will not
  // reflect change
  public static synchronized TiSession getSession(TiConfiguration conf) {
    String key = conf.getPdAddrsString();
    if (sessionCachedMap.containsKey(key)) {
      return sessionCachedMap.get(key);
    }

    TiSession newSession = TiSession.create(conf);
    sessionCachedMap.put(key, newSession);
    return newSession;
  }

  public static void clear() {
    sessionCachedMap.clear();
  }
}
